### PR TITLE
Better error messages

### PR DESCRIPTION
**Pull request by ivanprado**
* Grouping all errors found in the resultant exception.
* `full_final_kwargs` property in the plan.
* More polished code.

### CODE REVIEW

**Comment by kmike** [comment]
> Providing a list of all injection errors, instead just the first one. I think is Ok to fail just with the first as the failures are expected to happen at development time and fixing each case at each time seems reasonable to me. I don't think the complexity introduced is worth.

There is another use case when errors may happen: a set of externally_provided or is_injectable objects is changed after development of e.g. a scrapy-po page object is done. It may happen in production, not at development time, or at least not at development time of a scrapy-po object.

**Comment by ivanprado** [comment]
> There is another use case when errors may happen: a set of externally_provided or is_injectable objects is changed after development of e.g. a scrapy-po page object is done

I see. Working a little bit on that.

**Comment by ivanprado** [comment]
@kmike in the commit https://github.com/scrapinghub/andi/pull/13/commits/25f31ef608f3aa7afb3a4987fb115e4c490e0de6 I have changed the approach. Two main things:
* Now all the planning errors are accumulated and show to the user (see some examples below)
* There are two new functions to validate the plan: `assert_valid` and `assert_valid_and_complete`. Note that this means that always a function must be called after the planning, even in the "function" case. But we cannot save it because we need to validate in a single step to be able to accumulate all errors. 

Examples of errors shown:

```
Planning failed. Causes: 
0. Cyclic dependency found. Dependency graph: <class 'tests.test_plan.E'> -> <class 'tests.test_plan.C'> -> <class 'tests.test_plan.A'> -> <class 'tests.test_plan.E'>
1. Parameter 'e' in '<class 'tests.test_plan.A'>.__init__()' not resolvable due to underlying resolution errors.
2. Parameter 'a' in '<class 'tests.test_plan.C'>.__init__()' not resolvable due to underlying resolution errors.
3. Cyclic dependency found. Dependency graph: <class 'tests.test_plan.E'> -> <class 'tests.test_plan.D'> -> <class 'tests.test_plan.A'> -> <class 'tests.test_plan.E'>
4. Parameter 'e' in '<class 'tests.test_plan.A'>.__init__()' not resolvable due to underlying resolution errors.
5. Cyclic dependency found. Dependency graph: <class 'tests.test_plan.E'> -> <class 'tests.test_plan.D'> -> <class 'tests.test_plan.C'> -> <class 'tests.test_plan.A'> -> <class 'tests.test_plan.E'>
6. Parameter 'e' in '<class 'tests.test_plan.A'>.__init__()' not resolvable due to underlying resolution errors.
7. Parameter 'a' in '<class 'tests.test_plan.C'>.__init__()' not resolvable due to underlying resolution errors.
8. Parameter 'a' in '<class 'tests.test_plan.D'>.__init__()' not resolvable due to underlying resolution errors.
9. Parameter 'c' in '<class 'tests.test_plan.D'>.__init__()' not resolvable due to underlying resolution errors.
10. Parameter 'c' in '<class 'tests.test_plan.E'>.__init__()' not resolvable due to underlying resolution errors.
11. Parameter 'd' in '<class 'tests.test_plan.E'>.__init__()' not resolvable due to underlying resolution errors.
```

```
Planning failed. Causes: 
0. Any of [<class 'tests.test_plan.A'>] types are required  for parameter 'a'  in '<class 'tests.test_plan.C'>.__init__()' but none can be provided
1. Any of [<class 'tests.test_plan.B'>] types are required  for parameter 'b'  in '<class 'tests.test_plan.C'>.__init__()' but none can be provided
2. Parameter 'c' in '<class 'tests.test_plan.test_cannot_be_provided.<locals>.WithC'>.__init__()' not resolvable due to underlying resolution errors.
```

The work is uncomplete. Documentation must be updated, and probably some more tests would be required. But @kmike I prefer to wait until your feedback before working more on this direction.

**Comment by ivanprado** [comment]
After the changes for multiple-error message, I'm not sure anymore that is a good idea to have external validation just to avoid passing a parameter to the plan function. It seems pretty reasonable to have a `plan` parameter (e.g. `assert_complete`) that by default is False (the equivalent to the old parameter `full_final_kwargs`). 

Introducing `assert_valid` and `assert_valid_and_complete` seems more difficult to understand and it is more boilerplate.

**Comment by kmike** [comment]
Hey @ivanprado! Let's proceed with full_final_kwargs. Sorry for wasting so much of your time :(

**Comment by ivanprado** [comment]
@kmike please have a quick look. `full_final_kwargs` are back but with multi error support. Now errors are grouped by the arg name of the input function/class. Example errors:

```
Not possible to generate a plan for argument 'c' in '<class 'tests.test_plan.test_cannot_be_provided.<locals>.WithC'>.__init__()'. Causes:
    0. Any of [<class 'tests.test_plan.A'>] types are required for parameter 'a' in '<class 'tests.test_plan.C'>.__init__()' but none of them is injectable or externally providable
    1. Any of [<class 'tests.test_plan.B'>] types are required for parameter 'b' in '<class 'tests.test_plan.C'>.__init__()' but none of them is injectable or externally providable
```
```
Not possible to generate a plan for argument 'c' in '<class 'tests.test_plan.E'>.__init__()'. Causes:
    0. Cyclic dependency found. Dependency graph: <class 'tests.test_plan.E'> -> <class 'tests.test_plan.C'> -> <class 'tests.test_plan.A'> -> <class 'tests.test_plan.E'>
Not possible to generate a plan for argument 'd' in '<class 'tests.test_plan.E'>.__init__()'. Causes:
    0. Cyclic dependency found. Dependency graph: <class 'tests.test_plan.E'> -> <class 'tests.test_plan.D'> -> <class 'tests.test_plan.A'> -> <class 'tests.test_plan.E'>
    1. Cyclic dependency found. Dependency graph: <class 'tests.test_plan.E'> -> <class 'tests.test_plan.D'> -> <class 'tests.test_plan.C'> -> <class 'tests.test_plan.A'> -> <class 'tests.test_plan.E'>
```

```
Not possible to generate a plan for argument 'non_ann' in '<class 'tests.test_plan.test_plan_non_annotated_args.<locals>.WithNonAnnArgs'>.__init__()'. Causes:
    0. Parameter 'non_ann' is lacking annotations in '<class 'tests.test_plan.test_plan_non_annotated_args.<locals>.WithNonAnnArgs'>.__init__()'
Not possible to generate a plan for argument 'non_ann_def' in '<class 'tests.test_plan.test_plan_non_annotated_args.<locals>.WithNonAnnArgs'>.__init__()'. Causes:
    0. Parameter 'non_ann_def' is lacking annotations in '<class 'tests.test_plan.test_plan_non_annotated_args.<locals>.WithNonAnnArgs'>.__init__()'
Not possible to generate a plan for argument 'non_ann_kw' in '<class 'tests.test_plan.test_plan_non_annotated_args.<locals>.WithNonAnnArgs'>.__init__()'. Causes:
    0. Parameter 'non_ann_kw' is lacking annotations in '<class 'tests.test_plan.test_plan_non_annotated_args.<locals>.WithNonAnnArgs'>.__init__()'
Not possible to generate a plan for argument 'non_ann_kw_def' in '<class 'tests.test_plan.test_plan_non_annotated_args.<locals>.WithNonAnnArgs'>.__init__()'. Causes:
    0. Parameter 'non_ann_kw_def' is lacking annotations in '<class 'tests.test_plan.test_plan_non_annotated_args.<locals>.WithNonAnnArgs'>.__init__()'
```

Remaining work:

- [ ] Polish the code
- [ ] Change documentation
- [ ] Add tests for the `full_final_kwargs`

File: andi/andi.py [hunk 1]
```diff
@@ -106,10 +111,20 @@ def final_kwargs(self, instances: Mapping[Callable, Any]) -> Dict[str, Any]:
         return self[-1][1].kwargs(instances)
 
 
+    @property
```
Comment (kmike) [inline andi/andi.py]: nitpick: there are 2 blank lines before this property

File: andi/andi.py [hunk 2]
```diff
@@ -287,33 +303,49 @@ def _plan(class_or_func: Callable, *,
...
+        args_errs.update(non_injectable_errs)
+    flatten_errors = [error
+                      for errors in args_errs.values()
+                      for error in errors]
+    if is_root_call and args_errs:
+        _raise_exception(class_or_func, args_errs)
+
+    plan_od[class_or_func] = type_for_arg
+    plan = Plan(plan_od.items())
+    plan._full_final_kwargs = not non_injectable_errs
```
Comment (kmike) [inline andi/andi.py]: I think that's fine to remove the property, and just have a full_final_kwargs attribute without underscore.

File: andi/andi.py [hunk 3]
```diff
@@ -330,3 +362,18 @@ def _ensure_can_provide_func(cont_or_call: Optional[ContainerOrCallableType]
     if isinstance(cont_or_call, Container):
         return cont_or_call.__contains__
     return cont_or_call
+
+
+def _raise_exception(class_or_func, arg_errors):
```
Comment (kmike) [inline andi/andi.py]: A small suggestion: implement _exception_message function instead. It'd be easier to test, and it'd be more clear which exceptions the caller function raises.

File: andi/andi.py [hunk 4]
```diff
@@ -330,3 +362,18 @@ def _ensure_can_provide_func(cont_or_call: Optional[ContainerOrCallableType]
...
+    msg = ""
+    for idx, (arg, errors) in enumerate(arg_errors.items()):
+        if idx > 0:
+            msg += "\n"
+        msg += "Not possible to generate a plan for argument "
+        msg += "'{}' in '{}{}'. ".format(arg, class_or_func, init_str)
+        msg += "Causes:"
+        for idx, err in enumerate(errors):
+            msg += "\n    {}. {}".format(idx, err)
+    print(msg)
```
Comment (kmike) [inline andi/andi.py]: stray print

File: andi/andi.py [hunk 5]
```diff
@@ -287,33 +303,49 @@ def _plan(class_or_func: Callable, *,
...
+                msg += "but none of them is injectable or externally providable"
+            non_injectable_errs[argname].append(msg)
+
+    # Error managing
+    if full_final_kwargs:
+        args_errs.update(non_injectable_errs)
+    flatten_errors = [error
+                      for errors in args_errs.values()
+                      for error in errors]
+    if is_root_call and args_errs:
```
Comment (kmike) [inline andi/andi.py]: it seems this check can be done before building of flatten_errors

**Comment by kmike** [comment]
@ivanprado I took a quick look; haven't checked that errors work properly. Example output looks good. Left some comments, mostly related to "polish code" TODO item.

File: andi/andi.py (hunk 2 above)
Comment (ivanprado) [reply andi/andi.py]: Done

File: andi/andi.py (hunk 3 above)
Comment (ivanprado) [reply andi/andi.py]: Done

File: andi/andi.py (hunk 4 above)
Comment (ivanprado) [reply andi/andi.py]: Outch

File: andi/andi.py (hunk 5 above)
Comment (ivanprado) [reply andi/andi.py]: Changed

**Comment by ivanprado** [comment]
@kmike I have polished more the code. I think this is ready to be merged from my side, tell me what do you think.

**Comment by kmike** [comment]
Looks great, thanks @ivanprado!

I've merged it, but would you mind adding tests for the error message, as the logic to build it is now quite advanced?

**Comment by ivanprado** [comment]
> I've merged it, but would you mind adding tests for the error message, as the logic to build it is now quite advanced?

@kmike Sure, done at https://github.com/scrapinghub/andi/pull/14

**Pull request merged**